mlir/lib/Dialect/Rock/utility/transformMapUtils.cpp

```diff
@@ -1740,3 +1741,294 @@ Value mlir::rock::addPassThroughIndices(OpBuilder &b, Value transformed,
   }
   return ret;
 }
+
+enum DimType { Upper = 0, Lower = 1 };
+
+struct TransformAttrArgs {
+  rock::TransformType type;
+  std::pair<SmallVector<StringRef>, SmallVector<StringRef>> preservedNames;
+  std::pair<SmallVector<uint32_t>, SmallVector<uint32_t>> preservedDims;
+  SmallVector<int64_t> params;
+};
+
+template <DimType Type>
+SmallVector<uint32_t>
+getPreservedIndices(rock::TransformAttr tr,
```

Comment:
I think DimType could be an argument. no need for a template here.

Reply:
Yes, but it is not going to make a huge difference. I just wanted to stick to one style in this part of the code. I mean there are several auxiliary functions where I need `template <DimType Type>`. So, I decided to keep doing it.

Reply:
I think the templates, even if they're tripping our weird style sense, are reasonable.